File: python/second/non_overlapping_intervals.py
class Solution(object):
    # O(nlog(n)) time | O(n) space
    def eraseOverlapIntervals(self, intervals):
        """
        :type intervals: List[List[int]]
        :rtype: int
        """
        intervals.sort()

        prevEnd = intervals[0][1]
        res = 0
        for i in range(1, len(intervals)):
            start = intervals[i][0]
            end = intervals[i][1]
            if start < prevEnd:
                prevEnd = min(prevEnd, end)
                res += 1
            else:
                prevEnd = end
        return res


solution = Solution()
intervals = [[1,2],[2,3],[3,4],[1,3]]
print(solution.eraseOverlapIntervals(intervals))

File: src/main/java/june_2025/NonOverlappingIntervals.java
package june_2025;

import java.util.Arrays;

public class NonOverlappingIntervals {

    public static void main(String[] args) {
        int[][] intervals = {{1,2},{2,3},{3,4},{1,3}};
//        int[][] intervals = {{1,2},{1,2},{1,2}};

        NonOverlappingIntervals nonOverlappingIntervals = new NonOverlappingIntervals();
        int result = nonOverlappingIntervals.eraseOverlapIntervals(intervals);
        System.out.println(result);

    }

    // O(nlog(n)) time | O(n) space
    public int eraseOverlapIntervals(int[][] intervals) {
        Arrays.sort(intervals, (a, b) -> {
            if (a[0] - b[0] == 0) {
                return a[1] - b[1];
            }
            return a[0] - b[0];
        });

        int res = 0;
        int prevEnd = intervals[0][1];

        for (int i = 1; i < intervals.length; i++) {
            int start = intervals[i][0];
            int end = intervals[i][1];
            if (start >= prevEnd) {
                prevEnd = end;
            } else {
                res++;
                prevEnd = Math.min(prevEnd, end);
            }
        }

        return res;
    }


}
